**Actividad Integradora**

# 1 Importar Librería

In [165]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

import datetime


#2 Cargar base de datos

In [166]:
df = pd.read_excel('BD_Inmuebles_Limpieza_SUCIAMAS-2.xlsx')

df

Referencia Fecha Alta   Tipo Operación  Provincia Superficie  \
0            16 2004-01-05   Casa  Alquiler     Girona        179   
1            17 2004-01-05   Casa  Alquiler  Tarragona         58   
2            23 2004-01-10   Casa  Alquiler     Girona        183   
3            28 2004-01-12   Casa  Alquiler  Barcelona        187   
4            29 2004-01-12   Casa  Alquiler  Tarragona         91   
..          ...        ...    ...       ...        ...        ...   
495         452 2004-06-19  Suelo     Venta     Lleida        215   
496         453 2004-06-19  Suelo     Venta     Girona        185   
497         472 2004-06-25  Suelo     Venta  Tarragona        297   
498         474 2004-06-25  Suelo     Venta     Girona        199   
499         480 2004-06-28  Suelo     Venta     Lleida         58   

    Precio Venta Fecha Venta Vendedor Edad del vendedor  Numero cuartos  
0         937960  2004-11-21   Carmen                43             1.0  
1         358846  2004-10-08    Luisa                22             2.0  
2        1438929  2004-04-21    Luisa                70             1.0  
3        1660560  2004-06-16  Joaquín                55             NaN  
4         753571  2004-07-07    Luisa                26             2.0  
..           ...         ...      ...               ...             ...  
495      2097970  2004-11-13    Pedro                45             NaN  
496      1616900  2005-04-08    Jesús                35             NaN  
497      1820016  2005-01-08    María               NaN             2.0  
498      1622646  2005-06-23    Jesús               NaN             NaN  
499       530816  2004-10-07    Luisa                55             2.0  

[500 rows x 11 columns]

In [167]:
#Reiniciamos el index
df.reset_index()

index  Referencia Fecha Alta   Tipo Operación  Provincia Superficie  \
0        0          16 2004-01-05   Casa  Alquiler     Girona        179   
1        1          17 2004-01-05   Casa  Alquiler  Tarragona         58   
2        2          23 2004-01-10   Casa  Alquiler     Girona        183   
3        3          28 2004-01-12   Casa  Alquiler  Barcelona        187   
4        4          29 2004-01-12   Casa  Alquiler  Tarragona         91   
..     ...         ...        ...    ...       ...        ...        ...   
495    495         452 2004-06-19  Suelo     Venta     Lleida        215   
496    496         453 2004-06-19  Suelo     Venta     Girona        185   
497    497         472 2004-06-25  Suelo     Venta  Tarragona        297   
498    498         474 2004-06-25  Suelo     Venta     Girona        199   
499    499         480 2004-06-28  Suelo     Venta     Lleida         58   

    Precio Venta Fecha Venta Vendedor Edad del vendedor  Numero cuartos  
0         937960  2004-11-21   Carmen                43             1.0  
1         358846  2004-10-08    Luisa                22             2.0  
2        1438929  2004-04-21    Luisa                70             1.0  
3        1660560  2004-06-16  Joaquín                55             NaN  
4         753571  2004-07-07    Luisa                26             2.0  
..           ...         ...      ...               ...             ...  
495      2097970  2004-11-13    Pedro                45             NaN  
496      1616900  2005-04-08    Jesús                35             NaN  
497      1820016  2005-01-08    María               NaN             2.0  
498      1622646  2005-06-23    Jesús               NaN             NaN  
499       530816  2004-10-07    Luisa                55             2.0  

[500 rows x 12 columns]

In [168]:
#Cambiamos el nombre del index por Id
df.index.names = ['Id']

#Imprimimos el dataframe
df.head(10)

Referencia Fecha Alta  Tipo Operación  Provincia Superficie Precio Venta  \
Id                                                                             
0           16 2004-01-05  Casa  Alquiler     Girona        179       937960   
1           17 2004-01-05  Casa  Alquiler  Tarragona         58       358846   
2           23 2004-01-10  Casa  Alquiler     Girona        183      1438929   
3           28 2004-01-12  Casa  Alquiler  Barcelona        187      1660560   
4           29 2004-01-12  Casa  Alquiler  Tarragona         91       753571   
5           32 2004-01-13  Casa  Alquiler     Girona         92       627348   
6           70 2004-01-28  Casa  Alquiler        NaN        179      1647695   
7          103 2004-02-08  Casa  Alquiler     Girona        258      1188090   
8          122 2004-02-17  Casa  Alquiler  Barcelona        270      1385910   
9          150 2004-03-01  Casa  Alquiler     Lleida        212      1800516   

   Fecha Venta Vendedor Edad del vendedor  Numero cuartos  
Id                                                         
0   2004-11-21   Carmen                43             1.0  
1   2004-10-08    Luisa                22             2.0  
2   2004-04-21    Luisa                70             1.0  
3   2004-06-16  Joaquín                55             NaN  
4   2004-07-07    Luisa                26             2.0  
5   2004-04-12    María                12             2.0  
6   2004-07-12    María                44             NaN  
7   2004-07-19    María                35             2.0  
8   2005-02-05    Pedro                34             2.0  
9   2004-12-19    Jesús                22             2.0

# 3 Análisis exploratorio de datos (EDA)

Proceso:
*   Forma de la base de datos
*   Ver la cabeza y la cola
*   Tipos de datos
*   Estadística básica

La exploración de análisis también utiliza gráficas, por lo que este proceso lo haremos también después de la limpieza para comparar.






##Forma de la base de datos

In [169]:
#Usamos función shape para el total de renglones y columnas.
df.shape

(500, 11)

In [170]:
#Podemos ver solamente los renglones
len(df) #Renglones

500

In [171]:
#Podemos ver solamente las columnas
len(df.columns) #Columnas

11

##Ver los primeros (head) y los últimos (tail) registros
Este punto lo realizamos para una exploración rápida de los datos

In [172]:
# Primeros 5 renglones
df.head()

Referencia Fecha Alta  Tipo Operación  Provincia Superficie Precio Venta  \
Id                                                                             
0           16 2004-01-05  Casa  Alquiler     Girona        179       937960   
1           17 2004-01-05  Casa  Alquiler  Tarragona         58       358846   
2           23 2004-01-10  Casa  Alquiler     Girona        183      1438929   
3           28 2004-01-12  Casa  Alquiler  Barcelona        187      1660560   
4           29 2004-01-12  Casa  Alquiler  Tarragona         91       753571   

   Fecha Venta Vendedor Edad del vendedor  Numero cuartos  
Id                                                         
0   2004-11-21   Carmen                43             1.0  
1   2004-10-08    Luisa                22             2.0  
2   2004-04-21    Luisa                70             1.0  
3   2004-06-16  Joaquín                55             NaN  
4   2004-07-07    Luisa                26             2.0

In [173]:
# Últimos 5 renglones
df.tail()

Referencia Fecha Alta   Tipo Operación  Provincia Superficie  \
Id                                                                  
495         452 2004-06-19  Suelo     Venta     Lleida        215   
496         453 2004-06-19  Suelo     Venta     Girona        185   
497         472 2004-06-25  Suelo     Venta  Tarragona        297   
498         474 2004-06-25  Suelo     Venta     Girona        199   
499         480 2004-06-28  Suelo     Venta     Lleida         58   

    Precio Venta Fecha Venta Vendedor Edad del vendedor  Numero cuartos  
Id                                                                       
495      2097970  2004-11-13    Pedro                45             NaN  
496      1616900  2005-04-08    Jesús                35             NaN  
497      1820016  2005-01-08    María               NaN             2.0  
498      1622646  2005-06-23    Jesús               NaN             NaN  
499       530816  2004-10-07    Luisa                55             2.0

## Tipo de datos

Vemos que el tipo de datos sea correcto

In [174]:
df.dtypes

Referencia                    int64
Fecha Alta           datetime64[ns]
Tipo                         object
Operación                    object
Provincia                    object
Superficie                   object
Precio Venta                 object
Fecha Venta          datetime64[ns]
Vendedor                     object
Edad del vendedor            object
Numero cuartos              float64
dtype: object

In [175]:
#Vemos las columnas y su tipo correspondiente, en donde nos dice si se contiene nulos o no
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 11 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   Referencia         500 non-null    int64         
 1   Fecha Alta         500 non-null    datetime64[ns]
 2   Tipo               498 non-null    object        
 3   Operación          499 non-null    object        
 4   Provincia          492 non-null    object        
 5   Superficie         500 non-null    object        
 6   Precio Venta       500 non-null    object        
 7   Fecha Venta        500 non-null    datetime64[ns]
 8   Vendedor           499 non-null    object        
 9   Edad del vendedor  447 non-null    object        
 10  Numero cuartos     234 non-null    float64       
dtypes: datetime64[ns](2), float64(1), int64(1), object(7)
memory usage: 43.1+ KB


Como se puede observar, las columnas Superficie, Precio_venta y Edad_del_vendedor están marcados como objects cuando claramente son numéricos.

También se observa que tenemos columnas con valores nulos en Tipo, Operación, Provincia, Vendedor, Edad del vendedor y Numero cuartos

## Estadística básica

Obtenemos la estadística básica para ver la distribución de los datos

In [176]:
#Podemos ver la distribución de los datos
df.describe()

Referencia                  Fecha Alta                 Fecha Venta  \
count  500.000000                         500                         500   
mean   250.500000  2004-04-05 19:23:31.200000  2004-11-22 00:51:50.400000   
min      1.000000         2004-01-01 00:00:00         2004-04-05 00:00:00   
25%    125.750000         2004-02-19 00:00:00         2004-09-13 00:00:00   
50%    250.500000         2004-04-09 00:00:00         2004-11-17 00:00:00   
75%    375.250000         2004-05-22 06:00:00         2005-02-01 00:00:00   
max    500.000000         2004-07-04 00:00:00         2005-07-03 00:00:00   
std    144.481833                         NaN                         NaN   

       Numero cuartos  
count      234.000000  
mean         2.854701  
min          1.000000  
25%          2.000000  
50%          3.000000  
75%          4.000000  
max          6.000000  
std          1.483905

## Identificar nulos

In [177]:
#El total de nulos por columnas
total_of_all = df.isnull().sum()
total_of_all


Referencia             0
Fecha Alta             0
Tipo                   2
Operación              1
Provincia              8
Superficie             0
Precio Venta           0
Fecha Venta            0
Vendedor               1
Edad del vendedor     53
Numero cuartos       266
dtype: int64

In [178]:
#Total de porcentaje de nulos por columnas
percent_of_all = df.isnull().mean() # regresa la fracción de valores nulos
missing_data_test = pd.concat([total_of_all, percent_of_all*100], axis=1, keys=['Total Nulos', 'Percent (%)'])
missing_data_test

Total Nulos  Percent (%)
Referencia                   0          0.0
Fecha Alta                   0          0.0
Tipo                         2          0.4
Operación                    1          0.2
Provincia                    8          1.6
Superficie                   0          0.0
Precio Venta                 0          0.0
Fecha Venta                  0          0.0
Vendedor                     1          0.2
Edad del vendedor           53         10.6
Numero cuartos             266         53.2

## Adicionalmente usamos unique para ver los valores únicos de las columnas de interés

###Columna Tipo

In [179]:
df.Tipo.unique()

array(['Casa', 'Industrial', 'Local', nan, 'Oficina', 'Parking', 'Piso',
       'Pisar', 'Suelo'], dtype=object)

In [180]:
#Nos da el conteo de cada valor en orden descentente.
df.Tipo.value_counts()

#Otra manera de escribirlo
#df['Tipo'].value_counts()


Tipo
Suelo         82
Oficina       75
Industrial    74
Parking       74
Local         73
Casa          57
Piso          50
Pisar         13
Name: count, dtype: int64

La columna Tipo tiene el valor de "Pisar" que debe de cambiarse a Piso


###Columna Operación

In [181]:
df.Operación.unique()

array(['Alquiler', 'Venta', 'Alkiler', ' ', nan], dtype=object)

In [182]:
#Nos da el conteo de cada valor en orden descendente.
df.Operación.value_counts()

#Otra manera de escribirlo
#df['Operación'].value_counts()


Operación
Alquiler    249
Venta       243
Alkiler       6
              1
Name: count, dtype: int64

La columna Operación tiene escrita mal la palabra "Alkiler" y debe de cambiarse a Alquiler.

Tiene un valor en espacio.

###Columna Provincia

In [183]:
df.Provincia.unique()

array(['Girona', 'Tarragona', 'Barcelona', nan, 'Lleida', ' '],
      dtype=object)

In [184]:
#Nos da el conteo de cada valor en orden descentente.
df.Provincia.value_counts()

#Otra manera de escribirlo
#df['Provincia'].value_counts()


Provincia
Girona       135
Barcelona    122
Tarragona    121
Lleida       113
               1
Name: count, dtype: int64

La columna Provincia tiene un valor con espacio

###Columna Superficie

In [185]:
df.Superficie.unique()

array([179, 58, 183, 187, 91, 92, 258, 270, 212, 100, 87, 63, 56, 170,
       201, 181, 177, 295, 178, 298, 251, 149, 192, 72, 57, 143, 65, 281,
       278, 120, 235, 227, 130, 119, 284, 211, 156, 239, 204, 285, 229,
       146, 241, 141, 80, 123, 173, 90, 214, 155, 153, 261, 60, 61, 131,
       176, 142, 257, 140, 219, 112, 236, 93, 132, 103, 52, ' ', 193, 216,
       46, 209, 159, 150, 225, 277, 190, 242, 76, 247, 74, 287, 233, 248,
       268, 240, 228, 166, 54, 118, 194, 139, 88, 265, 163, 263, 226, 223,
       238, 106, 45, 196, 121, 189, 162, 243, 70, 206, 274, 68, 127, 49,
       125, 160, 231, 117, 169, 82, 182, 246, 237, 138, 259, 137, 203,
       260, 0, 44, 199, 40, 55, 109, 202, 291, 267, 230, 269, 224, 148,
       126, 215, 282, 79, 94, 124, 116, 275, 105, 134, 128, 174, 186, 244,
       110, 41, 207, 205, 77, 168, 144, 104, 253, 172, 42, 154, 86, 133,
       249, 102, 158, 107, 232, 69, 62, 217, 95, 152, 184, 296, 271, 234,
       299, 300, 51, 167, 164, 175, 200, 197, 20

In [186]:
#Nos da el conteo de cada valor en orden descentente.
df.Superficie.value_counts()

#Otra manera de escribirlo
#df['Superficie'].value_counts()


Superficie
       10
179     5
298     5
209     5
54      5
       ..
267     1
109     1
40      1
44      1
297     1
Name: count, Length: 228, dtype: int64

La columna Superficie tiene 10 valores con espacios

###Columna Precio Venta

In [187]:
df["Precio Venta"].unique()

array([937960, 358846, 1438929, 1660560, 753571, 627348, 1647695, 1188090,
       1385910, 1800516, 731700, 779868, 492156, 474600, 995520, 1107108,
       1449629, 924294, 1877675, 1458176, 2612566, 2439971, 1238339,
       1159296, 702576, 419178, 836836, 546780, 2483197, 1537062, 1140840,
       '.', 1467782, 1230190, 686154, 1561432, 875228, 537264, 1369680,
       1670965, 1176597, 1291116, 2534505, 1990697, 1083466, 2060550,
       732495, 354320, 775884, 1117753, 412740, 1962380, '$', 1506438,
       2475063, 253920, 531432, 953156, 820336, 1405374, 1799771, 1054340,
       948927, 1104992, 1812244, 778875, 723609, 922020, 965358, 728931,
       2067236, 508924, '%', 1658835, 1398600, 351256, 1042283, 1582527,
       968250, 712712, 1360800, 1171894, 2571111, 2456159, 977740,
       1227224, 2110966, 1263495, 501220, 1210053, 1331166, 2188138,
       729788, 2452989, 1541528, 1674744, 2628812, 2238000, 999328,
       2020992, 1457488, '¡', 1373816, 1142938, 509010, 1098055, 4387

In [188]:
#Nos da el conteo de cada valor en orden descentente.
df["Precio Venta"].value_counts()

#Otra manera de escribirlo
#df['Precio_venta'].value_counts()


Precio Venta
0          3
+          2
           2
$          2
*          2
          ..
1992228    1
1394461    1
877680     1
391391     1
530816     1
Name: count, Length: 494, dtype: int64

La columna Precio_venta tiene valores vacíos, ceros y múltiples simbolos que deben de ser corregidos

###Columna Vendedor

In [189]:
df.Vendedor.unique()

array(['Carmen', 'Luisa', 'Joaquín', 'María', 'Pedro', 'Jesús', nan],
      dtype=object)

In [190]:
#Nos da el conteo de cada valor en orden descentente.
df.Vendedor.value_counts()

#Otra manera de escribirlo
#df['Vendedor'].value_counts()


Vendedor
Luisa      100
Jesús       95
Pedro       93
Carmen      82
María       67
Joaquín     62
Name: count, dtype: int64

La columna vendedor tiene valores nulos

###Columna Edad del vendedor

In [191]:
df["Edad del vendedor"].unique()

array([43, 22, 70, 55, 26, 12, 44, 35, 34, 45, 66, nan, 56, 87, 57, 38,
       76, 20, 88, 224, 65, 27, 78, 46, ' ', 74, 33, 54, 67, 24, 77, 89,
       188], dtype=object)

In [192]:
#Nos da el conteo de cada valor en orden descentente.
df["Edad del vendedor"].value_counts()

#Otra manera de escribirlo
#df["Edad del vendedor"].value_counts()


Edad del vendedor
44     50
43     39
55     37
56     30
45     28
88     28
22     27
35     26
78     26
34     16
26     14
38     14
27     13
70     13
33     12
54      8
66      8
46      8
76      7
224     6
65      4
57      4
74      4
87      4
67      4
24      4
77      4
        3
12      3
89      1
20      1
188     1
Name: count, dtype: int64

La columna edad_vendedor tiene campos vacíos y edades que no entran dentro de un límite lógico como 224 y 12 años

###Columna Numero cuartos

In [193]:
df["Numero cuartos"].unique()

array([ 1.,  2., nan,  3.,  5.,  6.,  4.])

In [194]:
#Nos da el conteo de cada valor en orden descentente.
df["Numero cuartos"].value_counts()

#Otra manera de escribirlo
#df["Numero cuartos"].value_counts()


Numero cuartos
3.0    65
1.0    55
2.0    45
4.0    31
5.0    24
6.0    14
Name: count, dtype: int64

La columna Numero_cuartos tiene muchos campos vacíos

# 4 Preprocesamiento de los datos

Proceso:
*   Renombrar columnas (de ser necesario)
*   Eliminar columnas que no utilicemos
*   Eliminar las columnas que tengan alto porcentaje de valores nulos
*   Eliminar duplicados
*   Corregir el formato del tipo de datos
*   Obten una muestra de la base de datos (en caso de que la base de datos sea muy grande)
*   Elimina outliers y valores vacíos
*   Eliminar duplicados
*   Imputación de datos





## Renombrar columnas (de ser necesario)

In [195]:
df =df.set_axis(['Referencia','Fecha_alta','Tipo','Operación','Provincia',"Superficie", "Precio_venta", "Fecha_venta", "Vendedor", "Edad_vendedor","Numero_cuartos"], axis=1)
df
#Ahora se muestra el nuevo Data frame con los nombres por columna actualizados.

Referencia Fecha_alta   Tipo Operación  Provincia Superficie  \
Id                                                                  
0            16 2004-01-05   Casa  Alquiler     Girona        179   
1            17 2004-01-05   Casa  Alquiler  Tarragona         58   
2            23 2004-01-10   Casa  Alquiler     Girona        183   
3            28 2004-01-12   Casa  Alquiler  Barcelona        187   
4            29 2004-01-12   Casa  Alquiler  Tarragona         91   
..          ...        ...    ...       ...        ...        ...   
495         452 2004-06-19  Suelo     Venta     Lleida        215   
496         453 2004-06-19  Suelo     Venta     Girona        185   
497         472 2004-06-25  Suelo     Venta  Tarragona        297   
498         474 2004-06-25  Suelo     Venta     Girona        199   
499         480 2004-06-28  Suelo     Venta     Lleida         58   

    Precio_venta Fecha_venta Vendedor Edad_vendedor  Numero_cuartos  
Id                                                                   
0         937960  2004-11-21   Carmen            43             1.0  
1         358846  2004-10-08    Luisa            22             2.0  
2        1438929  2004-04-21    Luisa            70             1.0  
3        1660560  2004-06-16  Joaquín            55             NaN  
4         753571  2004-07-07    Luisa            26             2.0  
..           ...         ...      ...           ...             ...  
495      2097970  2004-11-13    Pedro            45             NaN  
496      1616900  2005-04-08    Jesús            35             NaN  
497      1820016  2005-01-08    María           NaN             2.0  
498      1622646  2005-06-23    Jesús           NaN             NaN  
499       530816  2004-10-07    Luisa            55             2.0  

[500 rows x 11 columns]

## Eliminar columnas que no utilicemos



In [196]:
df=df.drop(df.columns[[0,1]], axis = 1)
df.head()

Tipo Operación  Provincia Superficie Precio_venta Fecha_venta Vendedor  \
Id                                                                           
0   Casa  Alquiler     Girona        179       937960  2004-11-21   Carmen   
1   Casa  Alquiler  Tarragona         58       358846  2004-10-08    Luisa   
2   Casa  Alquiler     Girona        183      1438929  2004-04-21    Luisa   
3   Casa  Alquiler  Barcelona        187      1660560  2004-06-16  Joaquín   
4   Casa  Alquiler  Tarragona         91       753571  2004-07-07    Luisa   

   Edad_vendedor  Numero_cuartos  
Id                                
0             43             1.0  
1             22             2.0  
2             70             1.0  
3             55             NaN  
4             26             2.0

In [197]:
#Podemos usar del para eliminar columnas
#del df["Vendedor"]
df

Tipo Operación  Provincia Superficie Precio_venta Fecha_venta Vendedor  \
Id                                                                             
0     Casa  Alquiler     Girona        179       937960  2004-11-21   Carmen   
1     Casa  Alquiler  Tarragona         58       358846  2004-10-08    Luisa   
2     Casa  Alquiler     Girona        183      1438929  2004-04-21    Luisa   
3     Casa  Alquiler  Barcelona        187      1660560  2004-06-16  Joaquín   
4     Casa  Alquiler  Tarragona         91       753571  2004-07-07    Luisa   
..     ...       ...        ...        ...          ...         ...      ...   
495  Suelo     Venta     Lleida        215      2097970  2004-11-13    Pedro   
496  Suelo     Venta     Girona        185      1616900  2005-04-08    Jesús   
497  Suelo     Venta  Tarragona        297      1820016  2005-01-08    María   
498  Suelo     Venta     Girona        199      1622646  2005-06-23    Jesús   
499  Suelo     Venta     Lleida         58       530816  2004-10-07    Luisa   

    Edad_vendedor  Numero_cuartos  
Id                                 
0              43             1.0  
1              22             2.0  
2              70             1.0  
3              55             NaN  
4              26             2.0  
..            ...             ...  
495            45             NaN  
496            35             NaN  
497           NaN             2.0  
498           NaN             NaN  
499            55             2.0  

[500 rows x 9 columns]

## Eliminar las columnas que tengan alto porcentaje de valores nulos

El porcentaje que decidmos depende del contexto de nuestro problema.


In [198]:
#Ej. Borrar columnas con el 40 a 100% de nulos
df.head()

Tipo Operación  Provincia Superficie Precio_venta Fecha_venta Vendedor  \
Id                                                                           
0   Casa  Alquiler     Girona        179       937960  2004-11-21   Carmen   
1   Casa  Alquiler  Tarragona         58       358846  2004-10-08    Luisa   
2   Casa  Alquiler     Girona        183      1438929  2004-04-21    Luisa   
3   Casa  Alquiler  Barcelona        187      1660560  2004-06-16  Joaquín   
4   Casa  Alquiler  Tarragona         91       753571  2004-07-07    Luisa   

   Edad_vendedor  Numero_cuartos  
Id                                
0             43             1.0  
1             22             2.0  
2             70             1.0  
3             55             NaN  
4             26             2.0

In [199]:
#Eliminar columnas con el nombre
df.drop('Numero_cuartos', inplace=True, axis=1)

#Otra forma de eliminar columnas
#df.drop(columns='Numero_cuartos')

df

Tipo Operación  Provincia Superficie Precio_venta Fecha_venta Vendedor  \
Id                                                                             
0     Casa  Alquiler     Girona        179       937960  2004-11-21   Carmen   
1     Casa  Alquiler  Tarragona         58       358846  2004-10-08    Luisa   
2     Casa  Alquiler     Girona        183      1438929  2004-04-21    Luisa   
3     Casa  Alquiler  Barcelona        187      1660560  2004-06-16  Joaquín   
4     Casa  Alquiler  Tarragona         91       753571  2004-07-07    Luisa   
..     ...       ...        ...        ...          ...         ...      ...   
495  Suelo     Venta     Lleida        215      2097970  2004-11-13    Pedro   
496  Suelo     Venta     Girona        185      1616900  2005-04-08    Jesús   
497  Suelo     Venta  Tarragona        297      1820016  2005-01-08    María   
498  Suelo     Venta     Girona        199      1622646  2005-06-23    Jesús   
499  Suelo     Venta     Lleida         58       530816  2004-10-07    Luisa   

    Edad_vendedor  
Id                 
0              43  
1              22  
2              70  
3              55  
4              26  
..            ...  
495            45  
496            35  
497           NaN  
498           NaN  
499            55  

[500 rows x 8 columns]

## Eliminar duplicados

In [200]:
#Vemos los duplicados
df.duplicated()

Id
0      False
1      False
2      False
3      False
4      False
       ...  
495    False
496    False
497    False
498    False
499    False
Length: 500, dtype: bool

In [201]:
#Eliminar duplicados completa
df = df.drop_duplicates()
df

Tipo Operación  Provincia Superficie Precio_venta Fecha_venta Vendedor  \
Id                                                                             
0     Casa  Alquiler     Girona        179       937960  2004-11-21   Carmen   
1     Casa  Alquiler  Tarragona         58       358846  2004-10-08    Luisa   
2     Casa  Alquiler     Girona        183      1438929  2004-04-21    Luisa   
3     Casa  Alquiler  Barcelona        187      1660560  2004-06-16  Joaquín   
4     Casa  Alquiler  Tarragona         91       753571  2004-07-07    Luisa   
..     ...       ...        ...        ...          ...         ...      ...   
495  Suelo     Venta     Lleida        215      2097970  2004-11-13    Pedro   
496  Suelo     Venta     Girona        185      1616900  2005-04-08    Jesús   
497  Suelo     Venta  Tarragona        297      1820016  2005-01-08    María   
498  Suelo     Venta     Girona        199      1622646  2005-06-23    Jesús   
499  Suelo     Venta     Lleida         58       530816  2004-10-07    Luisa   

    Edad_vendedor  
Id                 
0              43  
1              22  
2              70  
3              55  
4              26  
..            ...  
495            45  
496            35  
497           NaN  
498           NaN  
499            55  

[500 rows x 8 columns]

In [202]:
#Eliminación de los registros duplicados parciales
df.duplicated(df.columns[~df.columns.isin(['Vendedor'])])

Id
0      False
1      False
2      False
3      False
4      False
       ...  
495    False
496    False
497    False
498    False
499    False
Length: 500, dtype: bool

## Obten una muestra de la base de datos (en caso de que la base de datos sea muy grande)

In [203]:
# Obten una muestra para trabajar de forma aleatoria

df.sample() #Nos da un solo renglón
df.sample(400) # Puedes usar directamente el número
df_sample=df.sample(n=400)
df_sample

Tipo Operación  Provincia Superficie Precio_venta Fecha_venta  \
Id                                                                      
17      Casa  Alquiler     Lleida        177       924294  2005-01-19   
223  Oficina  Alquiler  Tarragona        186      1382166  2005-03-01   
310  Parking  Alquiler     Lleida        263      1779458  2005-01-24   
432    Suelo  Alquiler  Tarragona        201      1874928  2004-06-21   
459    Suelo  Alquiler  Tarragona        110      1004740  2005-07-03   
..       ...       ...        ...        ...          ...         ...   
437    Suelo  Alquiler     Girona        136       873120  2004-12-26   
381     Piso  Alquiler     Lleida        114      1019160  2005-05-29   
463    Suelo     Venta  Tarragona        216      1478952  2004-10-15   
351  Parking     Venta  Barcelona        102       790602  2004-11-09   
203    Local     Venta     Lleida        282      1298328  2005-03-10   

    Vendedor Edad_vendedor  
Id                          
17     Luisa           NaN  
223    Pedro            44  
310    Pedro            78  
432    Jesús            45  
459    Luisa           NaN  
..       ...           ...  
437    Pedro            87  
381   Carmen            56  
463   Carmen            78  
351   Carmen           NaN  
203    Jesús            88  

[400 rows x 8 columns]

In [204]:
#Para generar una submuestra de tamaño mayor usamos el parametro replace = True
df.sample(n=5000, replace=True)


Tipo Operación  Provincia Superficie Precio_venta Fecha_venta  \
Id                                                                      
133    Local  Alquiler  Tarragona        201       939072  2004-08-04   
386     Piso  Alquiler  Tarragona        122      1093608  2005-02-27   
448    Suelo  Alquiler     Girona         75       683175  2004-09-04   
284  Parking  Alquiler  Tarragona        215      1154980  2005-01-14   
25      Casa  Alquiler  Tarragona         57       419178  2005-03-18   
..       ...       ...        ...        ...          ...         ...   
467    Suelo     Venta  Tarragona         96       625536  2004-11-07   
313  Parking  Alquiler  Barcelona        234      1911312  2005-04-08   
259  Oficina     Venta  Tarragona        117       739206  2005-01-25   
366     Piso  Alquiler     Lleida        250      2253500  2005-02-21   
318  Parking     Venta     Girona        117            +  2004-09-14   

    Vendedor Edad_vendedor  
Id                          
133   Carmen            27  
386    Luisa            35  
448  Joaquín            45  
284  Joaquín            22  
25     Jesús            56  
..       ...           ...  
467    Jesús            78  
313    Luisa            43  
259   Carmen            35  
366  Joaquín            45  
318    Pedro            55  

[5000 rows x 8 columns]

In [205]:
#Un porcentaje del total
df.sample(frac=0)

df.sample(frac=0.5).shape

df.sample(frac=1).shape

(500, 8)

In [206]:
# Primeros 1000 Renglones - Podemos usarlos para explorar las columnas y corregir el tipo de datos
df.head(300).to_csv('Actividad.csv')

In [207]:
# Ultimos 1000 Renglones - Podemos usarlos para explorar las columnas y corregir el tipo de datos
df.tail(300).to_csv('Actividad2.csv')

## Corregir el formato del tipo de datos


In [208]:
df.isnull().sum()

Tipo              2
Operación         1
Provincia         8
Superficie        0
Precio_venta      0
Fecha_venta       0
Vendedor          1
Edad_vendedor    53
dtype: int64

In [209]:
df.dtypes

Tipo                     object
Operación                object
Provincia                object
Superficie               object
Precio_venta             object
Fecha_venta      datetime64[ns]
Vendedor                 object
Edad_vendedor            object
dtype: object

### Cambiamos signos especiales usando pd.to_numeric coerce

Coerce convierte los valores no númericos a NaN



In [210]:
# Hacemos una copia para no modificar el dataframe original
df_coerce = df.copy()

# Usando pd.to_numeric, se cambian los errores a nulos
df_coerce['Superficie'] = pd.to_numeric(df_coerce['Superficie'], errors= 'coerce')
df_coerce['Precio_venta'] = pd.to_numeric(df_coerce['Precio_venta'], errors= 'coerce')
df_coerce['Edad_vendedor'] = pd.to_numeric(df_coerce['Edad_vendedor'], errors= 'coerce') #.isnull() nos va a decir si es verdadero o falso
df_coerce

Tipo Operación  Provincia  Superficie  Precio_venta Fecha_venta  \
Id                                                                      
0     Casa  Alquiler     Girona       179.0      937960.0  2004-11-21   
1     Casa  Alquiler  Tarragona        58.0      358846.0  2004-10-08   
2     Casa  Alquiler     Girona       183.0     1438929.0  2004-04-21   
3     Casa  Alquiler  Barcelona       187.0     1660560.0  2004-06-16   
4     Casa  Alquiler  Tarragona        91.0      753571.0  2004-07-07   
..     ...       ...        ...         ...           ...         ...   
495  Suelo     Venta     Lleida       215.0     2097970.0  2004-11-13   
496  Suelo     Venta     Girona       185.0     1616900.0  2005-04-08   
497  Suelo     Venta  Tarragona       297.0     1820016.0  2005-01-08   
498  Suelo     Venta     Girona       199.0     1622646.0  2005-06-23   
499  Suelo     Venta     Lleida        58.0      530816.0  2004-10-07   

    Vendedor  Edad_vendedor  
Id                           
0     Carmen           43.0  
1      Luisa           22.0  
2      Luisa           70.0  
3    Joaquín           55.0  
4      Luisa           26.0  
..       ...            ...  
495    Pedro           45.0  
496    Jesús           35.0  
497    María            NaN  
498    Jesús            NaN  
499    Luisa           55.0  

[500 rows x 8 columns]

In [211]:
df_coerce.isnull().sum()

Tipo              2
Operación         1
Provincia         8
Superficie       10
Precio_venta     16
Fecha_venta       0
Vendedor          1
Edad_vendedor    56
dtype: int64

In [212]:
df_coerce.dtypes

Tipo                     object
Operación                object
Provincia                object
Superficie              float64
Precio_venta            float64
Fecha_venta      datetime64[ns]
Vendedor                 object
Edad_vendedor           float64
dtype: object

### Cambiamos signos especiales usando astype

Es el comportamiento que utiliza to_numeric si no puede convertir un valor, raise hará una excepción, no cambia los valores no numéricos.

In [213]:
# Hacemos una copia para no modificar el dataframe original
df_raise = df.copy()

# Usando pd.to_numeric, se cambian los errores a nulos, nos vamos a encontrar con errores
#df_raise['Superficie'] = pd.to_numeric(df_raise['Superficie'], errors= 'raise')
#df_raise['Precio_venta'] = pd.to_numeric(df_raise['Precio_venta'], errors= 'raise')
#df_raise['Edad_vendedor'] = pd.to_numeric(df_raise['Edad_vendedor'], errors= 'raise')

##Si usaramos ignore, simplemente ignora la operación si encuentra un valor invalido

## Cambiar inconsistencia de variables

Por ejemplo, no puede exitir un tipo de vivienda que se llame pisar, por lo que tenemos que inferirir que se trata de piso, así que cambiaremos el nombre por piso

In [214]:
#Corregimos la variable usando replace
#Columna Tipo
df_coerce['Tipo']=df_coerce['Tipo'].replace('Pisar','Piso')

#Columna Operacion
df_coerce['Operación']=df_coerce['Operación'].replace(' ',0)

#Columna Superficie
df_coerce['Superficie']=df_coerce['Superficie'].replace(' ',0)

#Columna Precio_venta
df_coerce['Precio_venta']=df_coerce['Precio_venta'].replace(' ',0)

#Columna Edad_vendedor
df_coerce['Edad_vendedor']=df_coerce['Edad_vendedor'].replace('Pisar','Piso')


In [215]:
#Utilizamos de nuevo value_counts() para ver el nuevo conteo
df_coerce = df_coerce['Tipo'].value_counts()


## Eliminamos outliers y valores vacíos

###Valores vacíos

En este ejercicio vamos a observar que sucede si eliminamos todos los valores vacíos

In [216]:
df.isnull().sum()

Tipo              2
Operación         1
Provincia         8
Superficie        0
Precio_venta      0
Fecha_venta       0
Vendedor          1
Edad_vendedor    53
dtype: int64

In [217]:
#Eliminamos los valores NaNs de la tabla
df = df.dropna(axis=0)
df

Tipo Operación  Provincia Superficie Precio_venta Fecha_venta Vendedor  \
Id                                                                             
0     Casa  Alquiler     Girona        179       937960  2004-11-21   Carmen   
1     Casa  Alquiler  Tarragona         58       358846  2004-10-08    Luisa   
2     Casa  Alquiler     Girona        183      1438929  2004-04-21    Luisa   
3     Casa  Alquiler  Barcelona        187      1660560  2004-06-16  Joaquín   
4     Casa  Alquiler  Tarragona         91       753571  2004-07-07    Luisa   
..     ...       ...        ...        ...          ...         ...      ...   
493  Suelo     Venta  Tarragona        144       852480  2005-06-04   Carmen   
494  Suelo     Venta  Tarragona        105       783195  2004-11-20   Carmen   
495  Suelo     Venta     Lleida        215      2097970  2004-11-13    Pedro   
496  Suelo     Venta     Girona        185      1616900  2005-04-08    Jesús   
499  Suelo     Venta     Lleida         58       530816  2004-10-07    Luisa   

    Edad_vendedor  
Id                 
0              43  
1              22  
2              70  
3              55  
4              26  
..            ...  
493            22  
494            26  
495            45  
496            35  
499            55  

[437 rows x 8 columns]

In [218]:
#Reiniciamos el Index
df.dropna(axis=0).reset_index().tail(3)


Id   Tipo Operación Provincia Superficie Precio_venta Fecha_venta  \
434  495  Suelo     Venta    Lleida        215      2097970  2004-11-13   
435  496  Suelo     Venta    Girona        185      1616900  2005-04-08   
436  499  Suelo     Venta    Lleida         58       530816  2004-10-07   

    Vendedor Edad_vendedor  
434    Pedro            45  
435    Jesús            35  
436    Luisa            55

In [219]:
#Identificamos vacíos usando isna
df.isna().any()

Tipo             False
Operación        False
Provincia        False
Superficie       False
Precio_venta     False
Fecha_venta      False
Vendedor         False
Edad_vendedor    False
dtype: bool

In [220]:
#Identificamos vacíos usando isnull
df.isnull().sum()

Tipo             0
Operación        0
Provincia        0
Superficie       0
Precio_venta     0
Fecha_venta      0
Vendedor         0
Edad_vendedor    0
dtype: int64

## Resumen

En el notebook se identificaron varios problemas en la base de datos, entre ellos valores faltantes, registros duplicados, formatos inconsistentes y errores en la clasificación de categorías. Se observaron categorias numéricas almacenadas como texto y la presencia de símbolos no válidos que dificultaban el análisis. Para resolver estos problemas, se renombraron y eliminaron columnas innecesarias, y se corrigieron los tipos de datos utilizando funciones como pd.to_numeric y astype. Además, se eliminaron duplicados y outliers que podían sesgar los resultados. Estas acciones permitieron homogenizar la información, asegurando la vericidad de los datos y facilitando el análisis. El proceso realizado mejoró la calidad y confiabilidad de los resultados que se podrian obtener a partir de estos datos. En este caso no se utilizaron, pero se aplicarian técnicas de imputación para abordar los valores nulos, sustituyéndolos por medidas estadísticas como la media o la mediana, según el contexto. 